tests/cases/conformance/types/thisType/thisTypeInObjectLiterals2.ts

```diff
@@ -0,0 +1,144 @@
+// @declaration: true
+// @noImplicitAny: true
+// @noImplicitThis: true
+// @target: es5
+
+// In methods of an object literal with no contextual type, 'this' has the type
+// of the object literal.
+
+let obj1 = {
+    a: 1,
+    f() {
+        return this.a;
+    },
+    b: "hello",
+    c: {
+        g() {
+            this.g();
+        }
+    },
+    get d() {
+        return this.a;
+    },
+    get e() {
+        return this.b;
+    },
+    set e(value) {
+        this.b = value;
+    }
+};
+
+// In methods of an object literal with a contextual type, 'this' has the
+// contextual type.
```

Comment:
Yes, will do. However, the `backward` property in your example would be an error because it doesn't exist in the contextual type (due to excess property checking in #3823).

The excess property check may an irritant in cases where an object literal introduces additional properties that are used in methods of the object literal--although we haven't gotten any complaints about it so far and it's been there for a while (since #3823). We might consider allowing excess properties if those properties are actually witnessed in methods of the object literal, although that is a rather complex rule to implement and explain.